### ISSUE TITLE
About old version deprecation

### ISSUE THREAD

**Issue by miningpoolhub** [zcash/zcash#2822]
I recently found that my v1.0.11 zcash stopped working few hours ago due to deprecate version.
I updated to v1.0.13 now and it's working fine.

1) Is there anyway to know the deprecate reserved block height from zcash client?
I can't find the deprecate block height before it happens.

2) Or Is there any option to disable deprecation?
"disabledeprecation=1.0.11"
works but I want more general option like turning on/off
"disabledeprecation=1"

**Comment by str4d** [comment]
> Is there anyway to know the deprecate reserved block height from zcash client? I can't find the deprecate block height before it happens.

The current deprecation policy is 16 weeks (specifically, 64512 blocks) after a release, which means that you should always be on a supported release if you upgrade every three months. You can see the exact values for any particular version in the source file `src/deprecation.h`.

Two weeks before deprecation happens, you will see a warning both on the metrics screen and in your `debug.log`, listing the block height at which deprecation will occur. Search for the string `This version will be deprecated at block height`. If there is another place you would like to see these warnings, please let us know!

> Or Is there any option to disable deprecation? "disabledeprecation=1.0.11" works but I want more general option like turning on/off "disabledeprecation=1"

No, there is no way to permanently disable deprecation without modifying the source code. We specifically chose to require users to re-configure this setting every time they upgrade (if they want deprecation disabled), in order to ensure that users are thinking about what the upgrade is doing. We will be having whole-network upgrades in the coming months (first Overwinter in June, followed by Sapling in September), and the auto-deprecation (also called auto-senescence) feature helps us as developers to know that when a network upgrade happens, users are either already running the new software, or have explicitly opted to run old software (and thus opted out of the upgrade).

**Comment by miningpoolhub** [comment]
I don't think normal users would read debug.log
Maybe printing next deprecation block height in getinfo rpc call would be better.
Something like.
"deprecationheight":64512

Also as a pool operator, I don't like to update to every fresh latest release because I encountered bugs from releases many times. If it doesn't need hardfork, then wallet should be compatible and I want to stay in stable release.

Block explorer sites were down when deprecation happened. minergate, zchain.
Forced update is quite annoying if it doesn't provide auto update feature actually. Computer can do things in automatic but humans will make mistakes every 3~4 months here and there.

I hope you reconsider deprecation feature from normal user's perspective. They don't care performance improvements or wallet feature addition. They just want their wallets work continually.

**Comment by str4d** [comment]
> I don't think normal users would read debug.log. Maybe printing next deprecation block height in getinfo rpc call would be better. Something like. "deprecationheight":64512

Mmm, good idea. I think a dedicated RPC method would make the most sense (given that `getinfo` is deprecated and will likely disappear in future); I've opened #2828 for this. I've also opened #2829 to send out notifications via the configured `alertnotify`, which I expect most large node operators will be configuring.

> Also as a pool operator, I don't like to update to every fresh latest release because I encountered bugs from releases many times. If it doesn't need hardfork, then wallet should be compatible and I want to stay in stable release.

If upgrading once every three months is too often, then use the `disabledeprecation` option to stay on a particular version. Our network upgrades are hard forks though, so you would need to pay special attention to the roadmap in that case. See https://z.cash/blog/roadmap-update-2017-12.html for the latest update.

> Forced update is quite annoying if it doesn't provide auto update feature actually. Computer can do things in automatic but humans will make mistakes every 3~4 months here and there.

We considered this quite carefully, and decided that a deprecation approach was better for the Zcash network and its users than an auto-upgrade approach. See these two blog posts for our reasoning about release cycles and version lifetimes:

- https://z.cash/blog/release-cycle-and-lifetimes.html
- https://z.cash/blog/release-cycle-update.html
